File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        answer = [1] * n

        # Calculate prefix products
        for i in range(1, n):
            answer[i] = answer[i - 1] * nums[i - 1]

        # Calculate postfix products and multiply with prefix products
        postfix_prod = 1
        for i in range(n - 1, -1, -1):
            answer[i] *= postfix_prod
            postfix_prod *= nums[i]

        return answer

            
        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
    std::vector<int> answer(n, 1);

    // Calculate prefix products
    for (int i = 1; i < n; ++i) {
        answer[i] = answer[i - 1] * nums[i - 1];
    }

    // Calculate postfix products and multiply with prefix products
    int postfix_prod = 1;
    for (int i = n - 1; i >= 0; --i) {
        answer[i] *= postfix_prod;
        postfix_prod *= nums[i];
    }

    return answer;
        
    }
};